Add mechanism to turn any RAII guard into a Python Context Manager

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #102037

This PR:
- adds a mechanism to turn any RAII guard into a Python Context Manager
- turns ExcludeDispatchKeyGuard into a context manager, and purges usages
of the older torch._C.ExcludeDispatchKeyGuard from the codebase.

The mechanism is that given a RAII guard, we construct a context
manager object that holds an optional guard. When we enter the context
manager we populate the guard, when we exit we reset it.

We don't delete torch._C.ExcludeDispatchKeyGuard for BC reasons (people
are using it in fbcode). If this code actually sticks
(it is using C++17 and that worries me a bit), then I'll apply the
change to other RAII guards we have, otherwise, we can write our own
std::apply.

zou3519 committed:
Add mechanism to turn any RAII guard into a Python Context Manager

This PR:
- adds a mechanism to turn any RAII guard into a Python Context Manager
- turns ExcludeDispatchKeyGuard into a context manager, and purges usages
of the older torch._C.ExcludeDispatchKeyGuard from the codebase.

The mechanism is that given a RAII guard, we construct a context
manager object that holds an optional guard. When we enter the context
manager we populate the guard, when we exit we reset it.

We don't delete torch._C.ExcludeDispatchKeyGuard for BC reasons (people
are using it in fbcode). If this code actually sticks
(it is using C++17 and that worries me a bit), then I'll apply the
change to other RAII guards we have, otherwise, we can write our own
std::apply.

[ghstack-poisoned]

zou3519 committed:
Update on "Add mechanism to turn any RAII guard into a Python Context Manager"


This PR:
- adds a mechanism to turn any RAII guard into a Python Context Manager
- turns ExcludeDispatchKeyGuard into a context manager, and purges usages
of the older torch._C.ExcludeDispatchKeyGuard from the codebase.

The mechanism is that given a RAII guard, we construct a context
manager object that holds an optional guard. When we enter the context
manager we populate the guard, when we exit we reset it.

We don't delete torch._C.ExcludeDispatchKeyGuard for BC reasons (people
are using it in fbcode). If this code actually sticks
(it is using C++17 and that worries me a bit), then I'll apply the
change to other RAII guards we have, otherwise, we can write our own
std::apply.

[ghstack-poisoned]

zou3519 committed:
Update on "Add mechanism to turn any RAII guard into a Python Context Manager"


This PR:
- adds a mechanism to turn any RAII guard into a Python Context Manager
- turns ExcludeDispatchKeyGuard into a context manager, and purges usages
of the older torch._C.ExcludeDispatchKeyGuard from the codebase.

The mechanism is that given a RAII guard, we construct a context
manager object that holds an optional guard. When we enter the context
manager we populate the guard, when we exit we reset it.

We don't delete torch._C.ExcludeDispatchKeyGuard for BC reasons (people
are using it in fbcode). If this code actually sticks
(it is using C++17 and that worries me a bit), then I'll apply the
change to other RAII guards we have, otherwise, we can write our own
std::apply.

[ghstack-poisoned]

Skylion007 commented on torch/csrc/utils/python_raii.h:
```diff
@@ -0,0 +1,46 @@
+    auto emplace = [&](Args... args) {
+      return guard_.emplace(std::forward<Args>(args)...);
+    };
+    std::apply(emplace, args_);
```
```suggestion
    std::apply(std::move(emplace), args_);
```

zou3519 replied:
Does this require the captures to be movable as well?

zou3519 committed:
Update on "Add mechanism to turn any RAII guard into a Python Context Manager"


This PR:
- adds a mechanism to turn any RAII guard into a Python Context Manager
- turns ExcludeDispatchKeyGuard into a context manager, and purges usages
of the older torch._C.ExcludeDispatchKeyGuard from the codebase.

The mechanism is that given a RAII guard, we construct a context
manager object that holds an optional guard. When we enter the context
manager we populate the guard, when we exit we reset it.

We don't delete torch._C.ExcludeDispatchKeyGuard for BC reasons (people
are using it in fbcode). If this code actually sticks
(it is using C++17 and that worries me a bit), then I'll apply the
change to other RAII guards we have, otherwise, we can write our own
std::apply.

[ghstack-poisoned]

zou3519 committed:
Update on "Add mechanism to turn any RAII guard into a Python Context Manager"


This PR:
- adds a mechanism to turn any RAII guard into a Python Context Manager
- turns ExcludeDispatchKeyGuard into a context manager, and purges usages
of the older torch._C.ExcludeDispatchKeyGuard from the codebase.

The mechanism is that given a RAII guard, we construct a context
manager object that holds an optional guard. When we enter the context
manager we populate the guard, when we exit we reset it.

We don't delete torch._C.ExcludeDispatchKeyGuard for BC reasons (people
are using it in fbcode). If this code actually sticks
(it is using C++17 and that worries me a bit), then I'll apply the
change to other RAII guards we have, otherwise, we can write our own
std::apply.

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

bdhirsh reviewed (approved):
very cool!

bdhirsh reviewed (approved):
very cool!

zou3519 commented:
@pytorchbot merge -f "windows timed out"

swolchok commented on torch/csrc/utils/python_raii.h:
```diff
@@ -0,0 +1,46 @@
+
+  void enter() {
+    auto emplace = [&](Args... args) {
+      return guard_.emplace(std::forward<Args>(args)...);
```
strangely, this doesn't seem to compile when c10::optional is std::optional (#101995). haven't figured out exactly why yet.
